statsd/statsd.go

```diff
@@ -121,30 +135,100 @@ func (c *Client) watch() {
 
 func (c *Client) append(cmd string) error {
 	c.Lock()
+	defer c.Unlock()
 	c.commands = append(c.commands, cmd)
 	// if we should flush, lets do it
 	if len(c.commands) == c.bufferLength {
 		if err := c.flush(); err != nil {
-			c.Unlock()
 			return err
 		}
 	}
-	c.Unlock()
 	return nil
 }
 
+func joinMaxSize(cmds []string, sep string, maxSize int) ([]string, []int) {
+	var buffer bytes.Buffer
+	buffer.Reset() //just to be sure...
+
+	frames := make([]string, 0)
+	ncmds := make([]int, 0)
+	sepLength := len(sep)
+
+	elem := 0
+	for _, cmd := range cmds {
+		var needed int
+
+		if elem == 0 {
+			needed = len(cmd)
+		} else {
+			needed = sepLength + len(cmd)
+		}
+
+		if buffer.Len()+needed <= maxSize {
+			if elem != 0 {
+				buffer.WriteString(sep)
+			}
+			buffer.WriteString(cmd)
+			elem++
+		} else if elem == 0 {
+			//if a single command is too big, flush it and hope the UDP fragmented packet
+			//doesn't get dropped.
+			frames = append(frames, cmd)
+			ncmds = append(ncmds, 1)
+		} else {
+			frames = append(frames, buffer.String())
+			ncmds = append(ncmds, elem)
+			buffer.Reset()
+			buffer.WriteString(cmd)
+			elem = 1
```

Comment:
Shouldn't this be `elem = 0`? this case only happens when a new frame must be created and so we should clean it to 0.